### PR TITLE
DEV-451 | Custom 5xx page

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

N/A

#### What's this PR do?

-  Adds `500.html` — This page is a custom 500 error page that could be served by the Django app. It also should be uploaded to GCS, and the resulting URL should be included as a config var in Heroku (see below). 
- Adds `500_cloudflare.html` — This page contains everything in `500.html` plus some additional fields that Cloudflare wants to be in our custom error template.
- Adds a new url/view for the Cloudflare custom 500 error page. We expose this because we initially need to point Cloudflare at a live version of this page. It will scan that page and subsequently use the HTML in event of a 521 or 522 error (the latter of which is what originally led to DEV-451 being created as a ticket).
- Adds a dummy  url/view that is used to enable testing the `500.html` with a test client through the view layer.
- Cleans up `docs/deploy.md` —  I read this in order to piece together how GCS fit into everything, and noticed some formatting errors  and typos that I corrected. as well in case of a 500 error, but at this point, the URL configuration for the SPA swallows all

#### Why are we doing this? How does it help us?

When users encounter 500 error in the wild, they'll get an on-brand experience.

#### How should this be manually tested?

The custom 500.html page is only exposed in non-deploy environments. If you run the app locally, you can go to `localhost:8000/dummy-500-error` and you should get the custom 500 error page. The response status for that URL should be 500.  This page should look more or less like this:

Screen Shot 2022-06-21 at 8 47 49 PM

The custom `500_cloudflare.html` can be tested in a deployed environment (and locally) by going to `<domain>/cloudflare-500` -- if review app is unexpired, you can view here: https://rev-engine-dev-451-5-xx-xattj3.herokuapp.com/cloudflare-500 . The response status should be a 200, and you should see the custom 500 page. This page should look more or less like this:

Screen Shot 2022-06-21 at 8 46 00 PM

#### Have automated unit tests been added?

Yes.

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No 

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-541 | 5xx page](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-451)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

For Cloudflare, [follow these instructions](https://support.cloudflare.com/hc/en-us/articles/200172706-Customizing-Cloudflare-error-pages#4OLz0BwjruYzDeRhZyF3oB) to set up either an account- or domain-level custom error page.  After successfully publishing to Cloudflare, see next steps below.

For Heroku, [follow these instructions](https://devcenter.heroku.com/articles/error-pages#customize-pages) on customizing error pages. Specifically, we will need to upload `500.html` to somewhere hosting static assets — presumably, GCS — and then add the resulting URL as a config var in the bootstrap command as documented below in next steps.

#### Are there next steps? If so, what? Have tickets been opened for them?

[DEV-1954 | Clean up custom error view, url, test for Cloudflare after publishing](https://news-revenue-hub.atlassian.net/browse/DEV-1954)

[DEV-1955 Upload 500.html to GCS and add URL as ERROR_PAGE_URL to bootstrap-review-app.py](https://news-revenue-hub.atlassian.net/browse/DEV-1955)

### CODE REVIEW

File: docs/deploy.md [hunk 1]
```diff
@@ -1,87 +1,85 @@
...
+* [Heroku command line installed](https://devcenter.heroku.com/categories/command-line)
 * Docker
-* Access to the Git repository that will be used for heroku deploys.
-* Inclusion on a team that has billing access rights to the Heroku account.
-* [pwgen](https://github.com/jbernard/pwgen) or something like it.
-* A previously created Heroku app.
+* Access to the Git repository that will be used for Heroku deploys
+* Inclusion on a team that has billing access rights to the Heroku account
+* [pwgen](https://github.com/jbernard/pwgen) or something like it
+* A previously-created Heroku app
```
Comment (njharman) [inline docs/deploy.md:10]: Why do you hate punctuation?

File: docs/deploy.md [hunk 2]
```diff
@@ -1,87 +1,85 @@
...
+## Steps
 
-### Checkout the repository that will be used to deploy the app
+### 1. Checkout the repository that will be used to deploy the app
 
 ```shell
 $> git clone git@github.com:newsrevenuehub/rev-engine.git
 $> cd rev-engine
-(rev-engine) $> git checkout develop
+git checkout develop
```
Comment (njharman) [inline docs/deploy.md:19]: Why remove the $> prompt?  On this git command only?  Should be all removed, or all retained. +1 for removal when "better" ```shell formating available.

File: revengine/views.py [hunk 3]
```diff
@@ -91,3 +92,18 @@ def admin_select_options(request):
...
+    We manually point Cloudflare at the URL for this view, which will cause it to scane
+    this page and from then on it will display the scanned HTML for some 5xx errors.
+    For more info, see https://support.cloudflare.com/hc/en-us/articles/200172706-Configuring-Custom-Pages-Error-and-Challenge-
+    """
+    return render(request, "500_cloudflare.html", {})
+
+
+def dummy_view_for_raising_500(request):
+    """Used to simulate 500 errors"""
+    raise
```
Comment (njharman) [inline revengine/views.py:109]: Bare raise is weird. I'm surprised linters allow it. It's a runtime error so generates a 500. Still, weird.

I'd prefer explicit raise django.http500error(), if there is one. Or, "raise Exception()"

**Review by njharman — APPROVED** [review]
- Looks ok to me.  Added comments on suggestions for minor doc changes (the prompts, not the punctuation) and test quibble. 
- Tests pass,. 
- Got 500 on dummy url but not the html cause I have django-error bar or something similar intercepting uncaught exception and showing fancy traceback... that I don't want to figure out how to disable atm.

File: docs/deploy.md (hunk 2 above)
Comment (benjaminEwhite) [reply docs/deploy.md:19]: Oh, i meant to remove all of the `$>`s. I find them annoying, as it makes copy pasting more cumbersome.

File: docs/deploy.md (hunk 1 above)
Comment (benjaminEwhite) [reply docs/deploy.md:10]: Punctuation is for complete sentences within bullets, or bullets that together with an introductory phrase constitute a complete sentence.

https://apastyle.apa.org/style-grammar-guidelines/lists/bulleted

**Review by x110dc — APPROVED** [review]

**Pull request merged**